File: 2/main.cpp
#include <iostream>
#include <cmath>

#include <GLFW/glfw3.h>

void error_callback(int error, const char* description)
{
    std::clog << "Error: " << description << '\n';
}

static void key_callback(GLFWwindow* window, int key, int scancode, int action, int mods)
{
    if (key == GLFW_KEY_ESCAPE && action == GLFW_PRESS)
        glfwSetWindowShouldClose(window, GLFW_TRUE);
}

float deg_y = 0;

void drawCube() {
  glBegin(GL_QUADS);
  // face in xy plane
  glColor3f(0.82, 0.41,
            0.12); // this the color with which complete cube is drawn.
  glVertex3f(0, 0, 0);
  glVertex3f(1, 0, 0);
  glVertex3f(1, 1, 0);
  glVertex3f(0, 1, 0);

  // // face in yz plane
  // glColor3f(1, 0, 0);
  // glVertex3f(0, 0, 0);
  // glVertex3f(0, 0, 1);
  // glVertex3f(0, 1, 0);
  // glVertex3f(0, 1, 1);

  // // face in zx plance
  // glColor3f(0, 1, 0);
  // glVertex3f(0, 0, 0);
  // glVertex3f(0, 0, 1);
  // glVertex3f(1, 0, 1);
  // glVertex3f(1, 0, 0);

  // //|| to xy plane.
  // glColor3f(0, 0, 1);
  // glVertex3f(0, 0, 1);
  // glVertex3f(1, 0, 1);
  // glVertex3f(1, 1, 1);
  // glVertex3f(0, 1, 1);

  // //|| to yz plane
  // glColor3f(0.73, 0.58, 0.58);
  // glVertex3f(0, 0, 1);
  // glVertex3f(1, 0, 1);
  // glVertex3f(1, 1, 1);
  // glVertex3f(0, 1, 1);

  // //|| to zx plane
  // glVertex3f(0.58, 0, 0.82);
  // glVertex3f(0, 1, 0);
  // glVertex3f(0, 1, 1);
  // glVertex3f(1, 1, 1);
  // glVertex3f(1, 1, 0);
  glEnd();
}

void display() {
  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

  glMatrixMode(GL_PROJECTION);
  glLoadIdentity();
  glMatrixMode(GL_MODELVIEW);
  glLoadIdentity();
  // GLfloat model[4][4] = {
  //     {cosf(0.2f), 0, sinf(0.2f), 0},
  //     {0, 1, 0, 0},
  //     {-sinf(0.2f), 0, cosf(0.2f), 0},
  //     {0, 0, 0, 1},
  // };

  // glMultMatrixf(&model[0][0]);
  glTranslatef(-0.5, -0.5, 0);
  // glRotatef(30, 0, 0, 1);
  drawCube();


  // glPopMatrix();
}

int main(void) {
  GLFWwindow *window;

  glfwSetErrorCallback(error_callback);

  /* Initialize the library */
  if (!glfwInit())
    return -1;

  /* Create a windowed mode window and its OpenGL context */
  window = glfwCreateWindow(1920, 1080, "Lab2", NULL, NULL);
  if (!window) {
    glfwTerminate();
    return -1;
  }

  glfwSetKeyCallback(window, key_callback);

  /* Make the window's context current */
  glfwMakeContextCurrent(window);

  glfwSwapInterval(1);

  // Enable depth test
  glEnable(GL_DEPTH_TEST);
  // Accept fragment if it closer to the camera than the former one
  glDepthFunc(GL_LESS);

  glViewport(0, 0, 1920, 1080);

  /* Loop until the user closes the window */
  while (!glfwWindowShouldClose(window)) {
    display();

    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    glOrtho(0, 1920, 0, 1080, -1.0, 1.0);

    /* Swap front and back buffers */
    glfwSwapBuffers(window);

    /* Poll for and process events */
    glfwPollEvents();
  }

  glfwTerminate();
  return 0;
}

File: 2/main.py
#!/usr/bin/env python3

import glfw
from OpenGL.GL import *
from math import cos, sin, sqrt, asin

alpha = 0.0
beta = 0.0
fill = True

def main():
    if not glfw.init():
        return
    window = glfw.create_window(640, 640, "LAB 2", None, None)
    if not window:
        glfw.terminate()
        return
    glfw.make_context_current(window)
    glfw.set_key_callback(window, key_callback)

    glEnable(GL_DEPTH_TEST)
    glDepthFunc(GL_LESS)

    while not glfw.window_should_close(window):
        display(window)
    glfw.destroy_window(window)
    glfw.terminate()

def display(window):
    glLoadIdentity()
    glClear(GL_COLOR_BUFFER_BIT)
    glClear(GL_DEPTH_BUFFER_BIT)
    glMatrixMode(GL_PROJECTION)

    # move
    glMultMatrixf([1, 0, 0, 0,
                   0, 1, 0, 0,
                   0, 0, 1, 0,
                   0.75, 0.75, 0, 1])

    fz = 0
    theta = 0.61557763
    phi = 0.785398
    def projection():
        glMultMatrixf([
            cos(phi), sin(phi)*sin(theta), sin(theta)*cos(theta), 0,
            0, cos(theta), -sin(theta), 0,
            sin(phi), -cos(phi)*sin(theta), -cos(phi)*cos(theta), 0,
            0, 0, 0, 1,
        ])
    projection()

    def cube(sz):
        glBegin(GL_QUADS)
        glColor3f(0.0, 0.0, 1.0);
        glVertex3f(-sz/2, -sz/2, -sz/2)
        glVertex3f(-sz/2,  sz/2, -sz/2)
        glVertex3f(-sz/2,  sz/2,  sz/2)
        glVertex3f(-sz/2, -sz/2,  sz/2)
        glColor3f(1.0, 0.0, 0.0);
        glVertex3f( sz/2, -sz/2, -sz/2)
        glVertex3f( sz/2, -sz/2,  sz/2)
        glVertex3f( sz/2,  sz/2,  sz/2)
        glVertex3f( sz/2,  sz/2, -sz/2)
        glColor3f(0.0, 1.0, 0.0);
        glVertex3f(-sz/2, -sz/2, -sz/2)
        glVertex3f(-sz/2, -sz/2,  sz/2)
        glVertex3f( sz/2, -sz/2,  sz/2)
        glVertex3f( sz/2, -sz/2, -sz/2)
        glColor3f(1.0, 1.0, 0.0);
        glVertex3f(-sz/2, sz/2, -sz/2)
        glVertex3f(-sz/2, sz/2,  sz/2)
        glVertex3f( sz/2, sz/2,  sz/2)
        glVertex3f( sz/2, sz/2, -sz/2)
        glColor3f(0.0, 1.0, 1.0);
        glVertex3f(-sz/2, -sz/2, -sz/2)
        glVertex3f( sz/2, -sz/2, -sz/2)
        glVertex3f( sz/2,  sz/2, -sz/2)
        glVertex3f(-sz/2,  sz/2, -sz/2)
        glColor3f(1.0, 0.0, 1.0);
        glVertex3f(-sz/2, -sz/2,  sz/2)
        glVertex3f( sz/2, -sz/2,  sz/2)
        glVertex3f( sz/2,  sz/2,  sz/2)
        glVertex3f(-sz/2,  sz/2,  sz/2)
        glEnd()

    cube(0.2)

    glLoadIdentity()

    global alpha
    global beta
    x = 0.7
    y = 0.7
    fz = sqrt(x*x+y*y)
    theta = asin(fz/sqrt(2)) + alpha
    phi = asin(fz/sqrt(2-fz*fz)) + beta
    projection()

    cube(0.7)

    glfw.swap_buffers(window)
    glfw.poll_events()

def key_callback(window, key, scancode, action, mods):
    global alpha
    global beta
    if action == glfw.PRESS or action == glfw.REPEAT:
        if key == glfw.KEY_RIGHT:
            beta += 0.1
        elif key == glfw.KEY_LEFT:
            beta -= 0.1
        elif key == glfw.KEY_UP:
            alpha += 0.1
        elif key == glfw.KEY_DOWN:
            alpha -= 0.1
        elif key == glfw.KEY_F:
            global fill
            fill = not fill
            if fill:
                glPolygonMode(GL_FRONT_AND_BACK, GL_FILL);
            else:
                glPolygonMode(GL_FRONT_AND_BACK, GL_LINE);



if __name__ == "__main__":
    main()
